lib/rubygems/remote_fetcher.rb

```diff
@@ -21,19 +22,24 @@ class Gem::RemoteFetcher
   class FetchError < Gem::Exception
     ##
     # The URI which was being accessed when the exception happened.
+    def self.build(message, uri)
```

Comment:
Extracting the `initialize` method logic into the `build` class method and keeping the `initialize` method just for initializing the class.

Reply:
It feels a bit overkill to me to do this, specially given that it requires us to change all `FetchError` instantiations, and adapt the new `initialize` signature to still work with `bundler`. How about keep doing thins in `initialize`? If we also expose the original uri inside `Gem::PrintableUri`, the diff could be like this:

```diff
diff --git a/lib/rubygems/remote_fetcher.rb b/lib/rubygems/remote_fetcher.rb
index 8c9c9e6150..b9b1c3ac60 100644
--- a/lib/rubygems/remote_fetcher.rb
+++ b/lib/rubygems/remote_fetcher.rb
@@ -5,6 +5,7 @@
 require_relative 's3_uri_signer'
 require_relative 'uri_formatter'
 require_relative 'uri_parser'
+require_relative 'printable_uri'
 require_relative 'user_interaction'
 
 ##
@@ -25,15 +26,14 @@ class FetchError < Gem::Exception
     attr_accessor :uri, :original_uri
 
     def initialize(message, uri)
-      super message
-
-      uri = Gem::UriParser.parse_uri(uri)
+      uri = Gem::PrintableUri.parse_uri(uri)
 
-      @original_uri = uri.dup
+      message = message.sub(uri.original_password, 'REDACTED') if uri.original_password
 
-      uri.password = 'REDACTED' if uri.respond_to?(:password) && uri.password
+      super message
 
       @uri = uri.to_s
+      @original_uri = uri.original_uri.to_s
     end
 
     def to_s # :nodoc:
```